### DISCUSSION TITLE
How to get stake address from account pub key

### DISCUSSION THREAD

**Discussion by iTanChi** [bloxbean/cardano-client-lib#255] [Q&A]
I want to get the stake address from account pub key, which can be generated from Cardano wallet: eternl, ...
- The account pub key format can be one of: acct_vk, acct_xvk, xpub https://cips.cardano.org/cips/cip5/
The flow to generate I'm understating is:
- Get stake pub key from account pub key
- Get stake address from stake pub key

But currently I don't know how to use the lib to get stake pub key from account pub key. The only way I see is CIP1852

**Comment by satran004** [comment]
@iTanChi Here's the code to find stake address from an account pub key.

- Get account level HdPublicKey
- Generate stake key using correct derivation path.

```
@Test
    public void getStakeAddressFromAccountPublicKey() {
        String accountPubKeyBech32 = "acct_xvk136qnzfm6c34lddfxll60uqxwe3csymp8sdqvg3zcs79azchhqdn4qhs75qtwvuzjkd5h436fujcrgqgq2mnlmr5yse5zvewdj6flkgg5catgs";
        Bech32.Bech32Data decode = Bech32.decode(accountPubKeyBech32);
        byte[] data = decode.data;
        //Get account level HdPublicKey
        HdPublicKey accountPubKey = HdPublicKey.fromBytes(data);

        //Get stake key derivation path and derive stake key from account key. Stake key has role=2, index =0
        DerivationPath stakeDerivationPath = DerivationPath.createStakeAddressDerivationPath();
        HdKeyGenerator hdKeyGenerator = new HdKeyGenerator();
        HdPublicKey stakeRoleKey = hdKeyGenerator.getChildPublicKey(accountPubKey, stakeDerivationPath.getRole().getValue()); //or just use role as 2
        HdPublicKey stakeKey = hdKeyGenerator.getChildPublicKey(stakeRoleKey, stakeDerivationPath.getIndex().getValue());

        //Generate stake address from stake key
        Address stakeAddress = AddressProvider.getRewardAddress(stakeKey, Networks.testnet());
        System.out.println(stakeAddress.toBech32());

        Assertions.assertEquals("stake_test1uq28tn7xfckpaxxlfnhjz07r9nlwgcagzzza0r5la7hagycfa458m", stakeAddress.toBech32());

    }
```

I think we can add this as an utility method in the library. Can you please create an enhancement request on Github?

**Comment by satran004** [comment]
Added new methods in AddressProvider to get stake address from account public key. It will be available in the next release. (0.5.0)

https://github.com/bloxbean/cardano-client-lib/commit/2283c116965f2484d7f08e3feb6d49184190ffce
